csharp/src/Ice/LocalException.cs

```diff
@@ -85,51 +85,83 @@ public NoEndpointException(string stringifiedProxy)
         /// message.</summary>
         /// <param name="message">The message that describes the error.</param>
-        public TransportException(string message)
+        /// <param name="retryPolicy">The exception retry policy.</param>
+        /// <param name="connection">The connection associated with this failure.</param>
+        public TransportException(string message, RetryPolicy retryPolicy, Connection? connection = null)
             : base(message)
         {
+            RetryPolicy = retryPolicy;
+            Connection = connection;
         }
 
         /// <summary>Constructs a new instance of the <see cref="TransportException"/> class with a reference to the
         /// inner exception that is the cause of this exception.</summary>
         /// <param name="innerException">The exception that is the cause of the current exception.</param>
-        public TransportException(Exception innerException)
+        /// <param name="retryPolicy">The exception retry policy.</param>
+        /// <param name="connection">The connection associated with this failure.</param>
+        public TransportException(Exception innerException, RetryPolicy retryPolicy, Connection? connection = null)
             : base("", innerException)
         {
+            RetryPolicy = retryPolicy;
+            Connection = connection;
         }
 
         /// <summary>Constructs a new instance of the <see cref="LoadException"/> class with a specified error message
         /// and a reference to the inner exception that is the cause of this exception.</summary>
         /// <param name="message">The message that describes the error.</param>
         /// <param name="innerException">The exception that is the cause of the current exception.</param>
-        public TransportException(string message, Exception innerException)
+        /// <param name="retryPolicy">The exception retry policy.</param>
+        /// <param name="connection">The connection associated with this failure.</param>
+        public TransportException(
+            string message,
+            Exception innerException,
+            RetryPolicy retryPolicy,
+            Connection? connection = null)
             : base(message, innerException)
         {
+            RetryPolicy = retryPolicy;
+            Connection = connection;
         }
     }
 
     /// <summary>This exception reports a failed attempt to establish a connection.</summary>
     public class ConnectFailedException : TransportException
     {
         /// <summary>Constructs a new instance of the <see cref="ConnectFailedException"/> class.</summary>
-        public ConnectFailedException()
+        /// <param name="retryPolicy">The exception retry policy.</param>
+        /// <param name="connection">The connection associated with this failure.</param>
+        public ConnectFailedException(RetryPolicy retryPolicy, Connection? connection = null)
```

Comment:
Can we ever have a connection object when we failed to establish a connection?

Reply:
Yes, see https://github.com/zeroc-ice/ice/blob/e928491329a51b2994c665df2a4c6b2063b0e6e6/csharp/src/Ice/TcpTransceiver.cs#L64